src/Functions/factorial.cpp

```diff
@@ -0,0 +1,113 @@
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionUnaryArithmetic.h>
+#include <DataTypes/NumberTraits.h>
+#include <Common/FieldVisitorConvertToNumber.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+}
+
+template <typename A>
+struct FactorialImpl
+{
+    using ResultType = Int64;
```

Comment:
Why do we return `Int64` intead of `UInt64`, if function only returns values greater than 1 ?

Reply:
Because I wanted to keep result type same with Impala or Spark.

Reply:
You are right. I will modify it.

Reply:
@kitaisreal  done